node/src/Pos/Block/Logic/Header.hs

```diff
@@ -328,35 +327,86 @@ getHeadersOlderExp upto = do
+-- | Given optional @depthLimit@, @from@ and @to@ headers where @from@
+-- is older (not strict) than @to@, and valid chain in between can be
+-- found, headers in range @[from..to]@ will be found. If difference
+-- between headers is more than @depthLimit@, error will be thrown.
+getHeadersRange ::
+       forall ssc m. (HasConfiguration, MonadDBRead m, SscHelpersClass ssc)
+    => Maybe Word
+    -> HeaderHash
+    -> HeaderHash
+    -> m (Either Text (OldestFirst NE HeaderHash))
+getHeadersRange _ older newer | older == newer = runExceptT $ do
+    unlessM (isJust <$> (DB.blkGetHeader @ssc newer)) $
+        throwError "getHeadersRange: can't find newer-older header"
+    pure $ OldestFirst $ one newer
+getHeadersRange depthLimitM older newer = runExceptT $ do
     -- oldest and newest blocks do exist
-    start <- MaybeT $ DB.blkGetHeader @ssc newer
-    end   <- MaybeT $ DB.blkGetHeader @ssc older
-    guard $ getEpochOrSlot start >= getEpochOrSlot end
-    let lowerBound = getEpochOrSlot end
-    if newer == older
-    then pure $ one newer
-    else loadHeadersDo lowerBound (one newer) $ start ^. prevBlockL
+    newerHd <- fromMaybeM "can't retrieve newer header" $ DB.blkGetHeader @ssc newer
+    olderHd <- fromMaybeM "can't retrieve older header" $ DB.blkGetHeader @ssc older
+    let olderD = olderHd ^. difficultyL
+    let newerD = newerHd ^. difficultyL
+
+    -- Proving newerD >= olderD
+    let newerOlderF = "newer: "%build%", older: "%build
+    when (newerD == olderD) $
+        throwError $
+        sformat ("getHeadersRange: newer and older headers have "%
+                 "the same difficulty, but are not equal. "%newerOlderF)
+                newerHd olderHd
+    when (newerD < olderD) $
+        throwError $
+        sformat ("getHeadersRange: newer header is less dificult than older one. "%
+                 newerOlderF)
+                newerHd olderHd
+
+    -- How many epochs does this range cross.
+    let genDiff :: Int
+        genDiff = fromIntegral $ newerHd ^. epochIndexL - olderHd ^. epochIndexL
+    -- Number of blocks is difficulty difference + number of genesis blocks.
+    let depthDiff :: Word
+        depthDiff = fromIntegral $ genDiff + fromIntegral (newerD - olderD)
+
+    whenJust depthLimitM $ \depthLimit ->
+        when (depthDiff > depthLimit) $
+        throwError $
+        sformat ("getHeadersRange: requested "%int%" headers, but depthLimit is "%
+                 int%". Headers: "%newerOlderF)
+                depthDiff
+                depthLimit
+                newerHd olderHd
+
+    -- We load these depthDiff blocks.
+    let cond curHash _depth = curHash /= newer
+
+    -- This is [oldest..newest) headers, oldest first
+    allExceptNewest <- GS.loadHashesUpWhile older cond
+
+    -- Sometimes we will get an empty list, if we've just switched the
+    -- branch (after first checks are performed here) and olderHd is
+    -- no longer in the main chain.
```

Comment:
If you are going to make changes, you can put a reference to the issue you recently created about this piece of code. It's quite minor of course.